lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -1183,44 +1185,31 @@ private void onChannelMessage(ProtocolMessage message) {
     }
 
     private synchronized void onConnected(ProtocolMessage message) {
-        /* if the returned connection id differs from
-         * the existing connection id, then this means
-         * we need to suspend all existing attachments to
-         * the old connection.
-         * If realtime did not reply with an error, it
-         * signifies that this was a result of an earlier
-         * connection being invalidated due to being stale.
-         *
-         * Suspend all channels attached to the previous id;
-         * this will be reattached in setConnection() */
         ErrorInfo error = message.error;
-        if(connection.id != null && !message.connectionId.equals(connection.id)) {
-            /* we need to suspend the original connection */
-            if(error == null) {
-                error = REASON_SUSPENDED;
+
+        if (message.connectionId.equals(connection.id)) {
```

Comment:
Personally I think it's clearer to separate the resume logic explicitly:

```
connection.reason = message.error;
if (connection.id != null) { // there was a previous connection, so this is a resume and RTN15c applies
  if(message.connectionId.equals(connection.id)) {
    // resume succeeded
    if(message.error == null) {
      // RTN15c1: no action required wrt channel state
      Log.d(TAG, "connection has reconnected and resumed successfully");
    } else {
      // RTN15c2: no action required wrt channel state
      Log.d(TAG, "connection resume success with non-fatal error: " + error.message);
    }
    // send any messages still pending from the previous transport (RTN19a)
    sendPendingQueueMessages();
  } else {
    // RTN15c3: resume failed
    Log.d(TAG, "connection resume failed with error: " + error.message);
    channels.reattachOnResumeFailure();
    connection.connectionManager.msgSerial = 0;
    // send any messages still pending from the previous transport (RTN19a)
    // however, this time the re-sent pending messages have to have newly assigned `msgSerial`s. They can't simply be replayed, as they are in the successful resume case
    sendPendingQueueMessages();
  }
}
```

So perhaps `sendPendingQueueMessages()` needs a new boolean argument that tells it whether to reassign `msgSerial`s on the messages, starting from 0.

Reply:
This is where the if branch makes resume case explicit. 
https://github.com/ably/ably-java/pull/900/commits/15797488ef39486a780182ba5db97a13ca504860

https://github.com/ably/ably-java/pull/900/commits/15797488ef39486a780182ba5db97a13ca504860 adds new boolean parameter to reset start serial for pending messages